### PR TITLE
fix wrong streams.readLine (refs #9026)

### PR DESCRIPTION
**Pull request by narimiran**
`streams.readLine` is still very slow, but now at least it is correct (the same result as in Python or via `wc -l` or when you use `File` instead of `Stream`).

This is based on the solution by @rayman22201 from here: https://gist.github.com/rayman22201/6b26e1a14b5e4f04776103de976cd8d7#file-myreader-nim

### CODE REVIEW

**Comment by rayman22201** [comment]
this is cool. I'm glad it is correct now. That is important.
Note, this was not my implementation. (It is still the naive implementation from @iffy. A more correct naive implementation but still)
If you look here, my contribution was the `fgets` implementation:
https://gist.github.com/rayman22201/6b26e1a14b5e4f04776103de976cd8d7/revisions
and here:
https://gist.github.com/rayman22201/6b26e1a14b5e4f04776103de976cd8d7#file-fgetsreader-nim

My goal is that Streams should use `c_fgets` where possible. The performance problem as I understood it, is because using the `file` type directly uses the `fgets` implementation, but using it through a stream uses the naive implementation instead. `Socket` type has the same problem.

The `stream` type should defer to the more efficient implementation where possible.
My idea for accomplishing this was to add a readLineImpl to the Stream interface here: https://github.com/nim-lang/Nim/blob/master/lib/pure/streams.nim#L41

This would allow a form of dynamic dispatch, to get the most appropriate implementation...
I just never got around to implementing it.

File: tests/misc/tstreamsreadline.nim [hunk 1]
```diff
@@ -0,0 +1,27 @@
+discard """
```
Comment (timotheecour) [inline tests/misc/tstreamsreadline.nim]: the `discard` section here can be removed IMO since it doesn't test anything of interest; eg like `misc/tstrdist.nim`

File: tests/misc/tstreamsreadline.nim [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  output: ""
+"""
```
Comment (timotheecour) [inline tests/misc/tstreamsreadline.nim]: we have these, why not name it tests/stdlib/tstreams4.nim
```
tests/stdlib/tstreams.nim
tests/stdlib/tstreams2.nim
tests/stdlib/tstreams3.nim
```

File: tests/misc/tstreamsreadline.nim [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+let
+  filenames = ["bin.txt", "str.txt"]
+  expected = [14334, 20773]
+
+for i, filename in filenames:
+  var lines = 0
+  for x in readFile(newFileStream(filename)):
+    lines += 1
+  doAssert lines == expected[i]
```
Comment (timotheecour) [inline tests/misc/tstreamsreadline.nim]: also need to assert on content of 2 lines (eg 1st + last) to make sure line is exactly what we expect (makes sure it works for corner cases like line ending and end of file)

File: tests/misc/tstreamsreadline.nim (hunk 2 above)
Comment (narimiran) [reply tests/misc/tstreamsreadline.nim]: Because my search didn't yield those :) Fixed.

File: tests/stdlib/tstreams4.nim [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  while stream.readLine(line):
+    num += 1
+    yield line.len
+
+iterator readFile*(s: string): int =
+  for x in readFile(newStringStream(s)):
+    yield x
+
+let
+  filenames = ["bin.txt", "str.txt"]
```
Comment (timotheecour) [inline tests/stdlib/tstreams4.nim]: * instead of committing 2MB of random binary data to nim repo (it's not insignificant, really), why not simply auto-generate (with a fixed seed) random chars and assert that does what we expect? (and remove the file upon test completion)

* that function that generates text data could also be useful once we fix the 100x speed issue (https://github.com/nim-lang/Nim/issues/9975)
